nav2_controller/src/nav2_controller.cpp

```diff
@@ -279,6 +342,16 @@ void ControllerServer::computeControl()
       action_server_->terminate_current();
       return;
     }
+    
+    std::string gc_name ;
+    get_parameter("current_goal_checker", gc_name);
```

Comment:
Never get parameters at run time, use the stored default type. This logic also fails to communicate goal checker IDs from the action request to even utilize non-default types.

Reply:
Would be okay for you if I move this "goal checker update" to the parameters update callback and we check there if the requested goal checker exist?

Reply:
I don't think this is a very fragile way of handling multiple goal checkers that won't meet the needs of most uses of the feature. If you had, lets say 5 controllers, where 3 are typical point-to-point navigation path followers for specific situations, 1 is a docking controller, and 1 is a person following controller.

The 3 point to point navigation controllers would likely use the same goal checker, and the other 2 unique situations would use 2 unique goal checkers as well. In this case, the use of a specific goal checker is generally mapped to the use of a specific controller plugin. Changing the "parameter" to use would require an external agent to mark the new parameter outside of the context of the navigation task for each controller, which would be difficult.

Imagining another situation where the goal checker to use is not relative to the controller _only_ but as a component of it (maybe some situation where you want a rough and another precise) but still have a few different controller plugins. How would you know when to switch them and what that active controller is at the point you change the parameter? This external server that's changing the parameter would need a great deal of insight into the current state of the controller server and the behavior tree to time that correctly. If this logic occurred in the BT itself, that would make a little more sense, but not very sensible to change parameters then because you're inside the BT itself to set some information. 

So in light of that, can you highlight the situation you would like to change the goal checker during? If its mapped to specific controller IDs, then I think this is better served as a field in the controller namespace to select the name of the goal checker to use. If its changing based on some environmental conditions, then it should be baked into the controller update request.

Reply:
Hello I have been away for a couple of months.
I just noticed this answer you made in December.

Let me answer your points.

> So in light of that, can you highlight the situation you would like to change the goal checker during? If its mapped to specific controller IDs, then I think this is better served as a field in the controller namespace to select the name of the goal checker to use. If its changing based on some environmental conditions, then it should be baked into the controller update request.

In our cases we want/need to control the way the robot moves at any moment depending on the task the robot is doing. From my point of view, this decision is something that is made "application level", not at "navigation executive level". For our cases, navigation is a "service" we want to configure and use, but it is not the center of the application.

Answering to your specific question, IMHO, goal checkers could be reused or not depending on the case. they could even change for the same controller: I could need sometimes make more priority the precision in the goal meanwhile in others I could need to prioritize the speed. When I use different controllers probably the configuration would differ, but they could also be shared (after all, the parameters of the simple_goal_checker are quite simple conceptually)

Another possible alternative would just be being able to update at runtime "the goal checker" parameters, at least for the simple goal checker. 


> I think this is a very fragile way of handling multiple goal checkers that won't meet the needs of most uses of the feature. 

> Changing the "parameter" to use would require an external agent to mark the new parameter outside of the context of the navigation task for each controller, which would be difficult.

The solution we made has pros and cons. The discussion about the details can be long.
However, m point in this discussion is not how "this should be done". It is just a feature request: being able to configure externally the navigation stack .

> So in light of that, can you highlight the situation you would like to change the goal checker during? If its mapped to specific controller IDs, then I think this is better served as a field in the controller namespace to select the name of the goal checker to use. If its changing based on some environmental conditions, then it should be baked into the controller update request.

I am open to change the namespace if that is more convenient.

For our requirements, the "specific question" if the goal checker configuration is something that belong to the controller_node, the bt_navigator_node or to other node is not important, we can adapt. For example, for the navigation2 stack, the goal checking parameters were located in the local_planner, but in this fork we did we interact with the goal_checker of the bt_navigator node. All of these approaches may have cons and pros depending on the point of view and priorities of everyone. 

I've been trying to identify what is the most convenient approach from the authors of the nav2 stack point of view. But, to be sincere, I do not know that yet.

**In resume:** I have a requirement that I feel it cannot be easily satisfied by the nav2 at the current moment: "Being able to control the navigation configuration of the goal checking externally", also being able to change the "controller"  or the "planner". 

I would need to know what is the way I can configure the navigation stack in runtime externally. I do not understand very well why this feature was available in the ROS1 navigation stack and not now.

Nonetheless, I insist, at the current moment the fork ([ref](https://github.com/pabloinigoblasco/navigation2)) we have in the navigation2 stack work for our requirements. Maybe it does not fit on authors nav2 stack software design point of view, but at least it is something. 

How can we solve this issue?
I am open to any point of view to implement these features.

Reply:
Hi,

So its been awhile since I've looked at this so I've had a chance to look at it with fresh eyes. I think there's something we could do here very analog to the controller ID selection. So part of the [action definition](https://github.com/ros-planning/navigation2/blob/main/nav2_msgs/action/FollowPath.action#L3) we have the goal checker ID to use as well. Then if none is provided, it uses the single one if only a single one is around (else complain and reject the request like the controller does). That lets us have things like "precise" and "coarse" be high level BT navigator options that remaps down to specific goal checker instances with the finite specifics as to what "precise" and "coarse" means!! 

So you get the ability to have variable precision navigation goals and also simplify that from the perspective of the application/BT designer only working with the given goal checker IDs (like we do with controller IDs to abstract out the specific algorithms behind its role) 

---

So in responding to your comments: `In our cases we want/need to control the way the robot moves at any moment depending on the task the robot is doing.` since this is "task level", I think having this option as part of the action request directly meets that requirement! It also doesn't require a separate external server to set the parameter, now its just part of the behavior tree request (e.g. have subtrees for precise tasks and coarse tasks, or have a selector node for changing between them if there's not a clear behavior tree structured way of separating them -- this is application specific so there's no "right" answer).

---

> "Being able to control the navigation configuration of the goal checking externally", also being able to change the "controller" or the "planner".

We should chat about this more, can you PM me on Slack and we discuss this a little more synchronously? Or does this request described above fully fill your need?

I'd really like to be able to homologate your needs with the architecture here so that you don't need to have a separate fork to do what you need to do. Changing navigation request precision seems like a very reasonable feature we should find a way to better support. 

But changing the controller / planner algorithm is incredibly easy, all you do is change the `planner_id` or the `controller_id` port in your BT and "poof" a new algorithm is used. If you had different tasks, then your custom BT would be able to have calls to these BT nodes with different algorithms in their unique contexts!

Reply:
Hello Steve.
Thanks for your answer.

> So part of the action definition we have the goal checker ID to use as well. Then if none is provided, it uses the single one if only a single one is around (else complain and reject the request like the controller does). That lets us have things like "precise" and "coarse" be high level BT navigator options that remaps down to specific goal checker instances with the finite specifics as to what "precise" and "coarse" means!!

> So in responding to your comments: In our cases we want/need to control the way the robot moves at any moment depending on the task the robot is doing. since this is "task level", I think having this option as part of the action request directly meets that requirement! It also doesn't require a separate external server to set the parameter, now its just part of the behavior tree request (e.g. have subtrees for precise tasks and coarse tasks, or have a selector node for changing between them if there's not a clear behavior tree structured way of separating them -- this is application specific so there's no "right" answer).

IMHO that approach is convenient and easy to implement.

> We  should chat about this more, can you PM me on Slack and we discuss this a little more synchronously? Or does this request described above fully fill your need?

Sure, lets keep in contact via slack.

> I'd really like to be able to homologate your needs with the architecture here so that you don't need to have a separate fork to do what you need to do. Changing navigation request precision seems like a very reasonable feature we should find a way to better support.

That is great. It is not our goal to make forks from the navigation stack. It is just a way of progressing in some concepts or ideas. It is easier to talk if there is something done or there is some progress on the issue.

> But changing the controller / planner algorithm is incredibly easy, all you do is change the planner_id or the controller_id port in your BT and "poof" a new algorithm is used. If you had different tasks, then your custom BT would be able to have calls to these BT nodes with different algorithms in their unique contexts!

Yes, after some clarifications in some other git issues and the "experimental implementation I did" I think we are aligned here. It is creating a BTActionNode (or service). 

Altough 
In the experiment I  implemented in the fork was a decorator and parameters.(because I wanted to keep the life-cycle of the node for the whole execution). After some other discussions we had I a

Reply:
I think we covered the lifetime topic in slack, so I think this is mostly on the same page about adding this to the FollowPath action msg / BT port so that the same mechanics can apply to a selector node. You should be able to set from your external server the values for the inputs to the selector nodes via a topic / service!

The external node method is "legit" but I also want to offer both ways of doing things, for those that want to  build a little more reactive intelligence into their BT (so selector nodes for external servers to set, and ports for statically set algorithms in unique subtree contexts for people using conditionals to steer the conditions to set algorithms in context)